src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs

```diff
@@ -332,6 +332,48 @@ static async ValueTask<int> FinishReadAsync(Task<int> readTask, byte[] localBuff
             }
         }
 
+        public async ValueTask ReadExactlyAsync(Memory<byte> buffer, CancellationToken cancellationToken = default) =>
+            _ = await ReadAtLeastAsyncCore(buffer, buffer.Length, throwOnEndOfStream: true, cancellationToken).ConfigureAwait(false);
```

Comment:
This is going to force another state machine into existence for async waits, just to cast the `ValueTask<int>` to a `ValueTask`. We should probably change this to be more like:
```C#
public ValueTask ReadExactlyAsync(..)
{
    ValueTask<int> vt = ReadAtLeastAsyncCore(...);

    if (vt.IsCompletedSuccessfully)
        return default;

    return new ValueTask(vt.AsTask());
}
```
If ReadAtLeastAsyncCore doesn't end up being annotated to use a pooled builder, that shouldn't result in any additional allocation.  If it does end up using a pooled builder, that AsTask will allocate, which is unfortunate, and it'd be unfortunate to then try to work around that by making this method use a pooling builder as well... at which point the cheapest thing might be to just copy-and-paste the code from ReadAtLeastAsyncCore to here.  We unfortunately lack a general-purpose mechanism for freely converting any `ValueTask<T>` to a `ValueTask`.

Reply:
Actually... this is all in corelib.  We can cheat and use internals on `ValueTask<T>` to extract the backing object and use that to construct the `ValueTask`.

Reply:
I took a stab at writing it like this, but wasn't 100% sure I was doing it right. Let me know what you think.

Reply:
One question I have is do I need to access the Result of the underlying `ValueTask<int>` in order for it to be returned to the pool?

Reply:
> One question I have is do I need to access the Result of the underlying ValueTask<int> in order for it to be returned to the pool?

In general, yes.  But if you're transferring the internals of the `ValueTask<int>` over to the `ValueTask`, you definitely _don't_ want to do that.  It will be returned to the pool when the `ValueTask` is awaited by the consumer.